src/decay.hpp

```diff
@@ -0,0 +1,137 @@
+// along with FlexibleSUSY.  If not, see
+// <http://www.gnu.org/licenses/>.
+// ====================================================================
+
+#ifndef DECAY_H
+#define DECAY_H
+
+#include <initializer_list>
+#include <unordered_map>
+#include <vector>
+#include <string>
+
+#include <boost/core/demangle.hpp>
+
+namespace flexiblesusy {
+
+class Decay {
+public:
+   Decay(int, std::initializer_list<int>, double, std::string const&);
+   ~Decay() = default;
+   Decay(const Decay&) = default;
+   Decay(Decay&&) = default;
+   Decay& operator=(const Decay&) = default;
+   Decay& operator=(Decay&&) = default;
+
+   int get_initial_particle_id() const { return pid_in; }
+   const std::vector<int>& get_final_state_particle_ids() const {
+      return pids_out;
+   }
+   std::size_t get_final_state_size() const { return pids_out.size(); }
+
+   double get_width() const { return width; }
+   std::string get_proc_string() const { return proc_string; }
+   void set_width(double w) { width = w; }
+
+private:
+   int pid_in{0};
+   std::vector<int> pids_out{};
+   double width{0.};
+   std::string proc_string;
+};
+
+std::size_t hash_decay(const Decay& decay);
+
+class Decays_list {
+private:
+   using List_type = std::unordered_map<std::size_t, Decay>;
+public:
+   using iterator = List_type::iterator;
+   using const_iterator = List_type::const_iterator;
+
+   explicit Decays_list(int);
+   ~Decays_list() = default;
+   Decays_list(const Decays_list&) = default;
+   Decays_list(Decays_list&&) = default;
+   Decays_list& operator=(const Decays_list&) = default;
+   Decays_list& operator=(Decays_list&&) = default;
+
+   iterator begin() noexcept { return decays.begin(); }
+   const_iterator begin() const noexcept { return decays.begin(); }
+   const_iterator cbegin() const noexcept { return decays.cbegin(); }
+   iterator end() noexcept { return decays.end(); }
+   const_iterator end() const noexcept { return decays.end(); }
+   const_iterator cend() const noexcept { return decays.end(); }
+
+   std::size_t size() const noexcept { return decays.size(); }
+
+   void clear();
+   // @todo
+   // the string cannot be taken by const&
+   // if you do this, the printed to the screen total_width of Higgs boson
+   // in the SM is shifted from 4.07 MeV 3.98 MeV
+   // none of this makes any sense
+   // this happend with clang 10.0.0, but not with gcc 9.3.1
+   void set_decay(double width, std::initializer_list<int> products, std::string);
+   int get_particle_id() const { return initial_pdg; }
+   const Decay& get_decay(std::initializer_list<int> products) const;
+   double get_total_width() const { return total_width; }
+
+private:
+   int initial_pdg{0};
+   std::unordered_map<std::size_t, Decay> decays{};
+   double total_width{0.};
+};
+
+template<typename FieldIn, typename FieldOut1, typename FieldOut2>
+std::string wrong_width_sign_msg(std::vector<int> in, std::vector<int> out1, std::vector<int> out2) {
```

Comment:
I would strongly vote for removing this function and simply generate custom strings in the meta code. The calls to `set_decay()` would then look like this:
~~~
decays.set_decay(..., "hh -> {hh, hh}");
~~~
This would have the following advantages:
* Passing a string literal to `set_decay()` has no run-time overhead, because it is simply a `const char* const`.
* No custom `std::string` is created *in each call*, which may be very expensive as `set_decay()` is called many times. (The string to generate is known at meta code time.)
* The dependency on `boost::core::demangle` could be removed.
* Code bloat is reduced, because `wrong_width_sign_msg<A,B,C>` is instantiated *many* times, namely `|A|*|B|*|C|` times.

Reply:
It's a shame. It's such a fancy function :) I would still have to pass `std::array` as I want to add indices to an error message in a nice way (an information that is not available at the meta level). Is that ok with you?